src/engine/strat_engine/metadata.rs

```diff
@@ -4,262 +4,454 @@
+            1 => 0,
+            _ => panic!("invalid val from older()"),
         }
+    }
+}
 
-        let pool_uuid = try!(Uuid::parse_str(from_utf8(&buf[offset + 32..offset + 64]).unwrap())
-            .map_err(|_| "invalid pool uuid"));
-        let dev_uuid = try!(Uuid::parse_str(from_utf8(&buf[offset + 64..offset + 96]).unwrap())
-            .map_err(|_| "invalid dev uuid"));
+#[derive(Debug, Clone, Copy)]
+pub struct MDAHeader {
+    pub last_updated: Option<Timespec>,
+    pub used: u64,
+    pub region_size: u64,
+    pub data_crc: u32,
+}
+
+impl MDAHeader {
+    pub fn new(region_size: u64) -> MDAHeader {
+        MDAHeader {
+            last_updated: None,
+            used: 0,
+            region_size: region_size,
+            data_crc: 0,
+        }
+    }
 
-        let mda_size = Sectors(LittleEndian::read_u32(&buf[offset + 160..offset + 164]) as u64);
+    pub fn from_buf(buf: &[u8; MDA_REGION_HDR_SIZE], region_size: u64) -> EngineResult<MDAHeader> {
+        if LittleEndian::read_u32(&buf[..4]) != crc32::checksum_ieee(&buf[4..]) {
+            return Err(EngineError::Engine(ErrorEnum::Invalid, "MDA region header CRC".into()));
+        }
 
-        match validate_mda_size(mda_size) {
-            None => {}
-            Some(err) => {
-                return Err(err.into());
+        let time = {
+            match LittleEndian::read_i64(&buf[16..24]) {
+                0 => None,
+                secs => Some(Timespec::new(secs, LittleEndian::read_i32(&buf[24..28]))),
             }
         };
 
-        Ok(SigBlock {
-            mda: MDAGroup::read(buf, offset + 96, mda_size),
-            pool_uuid: pool_uuid,
-            dev_uuid: dev_uuid,
-            total_size: total_size,
-            mda_sectors: mda_size,
-            reserved_sectors:
-                Sectors(LittleEndian::read_u32(&buf[offset + 164..offset + 168]) as u64),
+        Ok(MDAHeader {
+            used: LittleEndian::read_u64(&buf[8..16]),
+            last_updated: time,
+            region_size: region_size,
+            data_crc: LittleEndian::read_u32(&buf[4..8]),
         })
     }
 
-    /// Read from Stratis magic number area in buf at offset.
-    pub fn read_strat_magic(buf: &[u8], offset: usize) -> &[u8] {
-        &buf[offset + 4..offset + 20]
-    }
+    pub fn to_buf(data: &[u8], timestamp: &Timespec) -> [u8; MDA_REGION_HDR_SIZE] {
+        let mut buf = [0u8; MDA_REGION_HDR_SIZE];
 
-    /// Read pool uuid from buf at offset.
-    /// Return an error if the data can't be interpreted as a Uuid.
-    pub fn read_pool_uuid(buf: &[u8], offset: usize) -> Result<Uuid, String> {
-        Uuid::parse_str(from_utf8(&buf[offset + 32..offset + 64]).unwrap())
-            .map_err(|_| "invalid pool uuid".into())
+        LittleEndian::write_u32(&mut buf[4..8], crc32::checksum_ieee(data));
+        LittleEndian::write_u64(&mut buf[8..16], data.len() as u64);
+        LittleEndian::write_u64(&mut buf[16..24], timestamp.sec as u64);
```

Comment:
Then the read methods should be changed to unsigned as well, so the operations are inverses of each other, as:

```
let time = {
    match LittleEndian::read_u64(&buf[62..24]) {
      0 => None,
      # signed cast are always safe, as highest order bit of each value read is guaranteed to be 0.
      secs => Some(Timespec::new(secs as i64, LittleEndian::read_u32(&buf[24..28]) as i32),
    }
};
```

It would be nice to comment writes as well, as:

```
# unsigned casts are always safe, as sec and nsec values are never negative
```

Otherwise it looks too much like a bug.